### ISSUE TITLE
Clarification of IAsyncQueryProvider

### ISSUE THREAD

**Issue by axelheer** [dotnet/efcore#2900]
I'm confused about the new interface `IAsyncQueryProvider` compared to the old one (`IDbAsyncQueryProvider`). Apparently the non generic stuff has been dropped, the compliance with the BCL naming guideline of appending Async to async methods as well (`IAsyncEnumerable<T>`, `IAsyncEnumerator<T>`). Furthermore there are actually two interfaces, `IAsyncQueryProvider` and `IEntityQueryProvider`. :confused:

So far so good, but I'm fiddling around with the fact, that the new query provider has two different execute methods, which I find a bit unlikely compared to other query providers I know.

```
public interface IQueryProvider
{
    IQueryable CreateQuery(Expression expression);
    IQueryable<TElement> CreateQuery<TElement>(Expression expression);
    object Execute(Expression expression);
    TResult Execute<TResult>(Expression expression);
}

public interface IAsyncQueryProvider : IQueryProvider
{
    IAsyncEnumerable<TResult> ExecuteAsync<TResult>(Expression expression);
    Task<TResult> ExecuteAsync<TResult>(Expression expression, CancellationToken cancellationToken);
}

public interface IEntityQueryProvider : IAsyncQueryProvider
{
}
```

Shouldn't be the result of `CreateQuery` _the_ enumerable, whether it's used for further query chaining or final enumeration? Now `CreateQuery` creates an `EntityQueryable` which calls back to `EntityQueryProvider` for async enumeration and only for async enumeration. Okay, it's not hard to follow, but it seems a bit odd in the first place.

I'm asking because I'm working on a [little tool](https://github.com/axelheer/nein-linq), which needs to intercept LINQ queries. The ["proxy" for async EF6 queries](https://github.com/axelheer/nein-linq/blob/master/src/NeinLinq.EF6/RewriteQueryProvider.cs) was quite straight forward to implement, the [version for EF7](https://github.com/axelheer/nein-linq/blob/master/src/NeinLinq.EF7/RewriteQueryProvider.cs) doesn't look right since it needs an extra enumerable dummy to fake async calls within mocking scenarios.

Maybe opening an issue isn't the right way to bring this into question, but since EF7 isn't carved into stone yet I thought it might not be that bad either. :innocent:

**Comment by anpete** [comment]
Hi @axelheer,

Let me try and address these in turn:

1) WRT the non-generic stuff being dropped, we haven't needed it directly and have a strong preference for the strongly-typed APIs. Feel free to file an issue for this specifically and we can consider adding it.

2) Regarding the Async suffix, observe that this interface actually lives in one of our dependencies: Ix-Async. We are still using the suffix on any top-level APIs that we own.

3) IEntityQueryProvider was introduced by @ajcvickers as a DI service contract. Given that it's empty, we may be able to just use IAsyncQueryProvider directly.

4) Re CreateQuery, it was a while since I did this, but IIRC I had to jump through a few hoops to integrate async querying with re-linq (another of our query dependencies) - re-linq only handles sync queries at this point. Happy to look at any PR which improves the situation.

Hope this helps,
Andrew.

**Comment by axelheer** [comment]
Thanks for the elaborate response!

It's a bit tedious -- every library / framework that builds on top of LINQ does its own async stuff, if at all. It would be nice to have some "standard" async LINQ, but I think it's to late for this anyway. :disappointed:

ad 1) I prefer strongly-typed APIs too, was just wondering since the "ordinary" LINQ API provides / requires both. Personally, I don't use the non-generic stuff either.

ad 2) Ah, I didn't realize that. It's still not nice, but too late to fix this...

ad 3) Okay, seems legit. It's a bit confusing though, if you want to plug something in: should I implement `IAsyncQueryProvider` or `IEntityQueryProvider`? At first sight `IAsyncQueryProvider` seems to work, but it's not just obvious. At least some documentation would help.

ad 4) Ah, dependency hell. I can play a bit around with this stuff, but I'm probable not able to improve this, so don't expect anything soon from me.

**Comment by HappyNomad** [comment]
@axelheer 

> It's a bit tedious -- every library / framework that builds on top of LINQ does its own async stuff, if at all. It would be nice to have some "standard" async LINQ

It looks like `System.Interactive.Async` (Ix) has taken on this role.  I just [requested a feature](https://github.com/Reactive-Extensions/Rx.NET/issues/294) to bridge a remaining gap with EF-Core.

**Comment by axelheer** [comment]
Oh no, `IAsyncQueryable` does not implement `IQueryable`. Thus, everything has to be done twice... 😞
